registry/client/repository.go

```diff
@@ -163,67 +158,156 @@ func (r *repository) Signatures() distribution.SignatureService {
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 
-	resp, err := ms.client.Get(u)
+	resp, err := t.client.Get(u)
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 	defer resp.Body.Close()
 
 	if SuccessStatus(resp.StatusCode) {
 		b, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return nil, err
+			return 0, err
 		}
 
 		tagsResponse := struct {
 			Tags []string `json:"tags"`
 		}{}
 		if err := json.Unmarshal(b, &tagsResponse); err != nil {
-			return nil, err
+			return 0, err
 		}
 
-		return tagsResponse.Tags, nil
+		copy(refs, tagsResponse.Tags)
+		if len(tagsResponse.Tags) > len(refs) {
+			// passed in refs was insufficient to capture all tags
+			return len(refs), nil
+		}
+		return len(tagsResponse.Tags), io.EOF
 	}
-	return nil, handleErrorResponse(resp)
+	return 0, handleErrorResponse(resp)
 }
 
-func (ms *manifests) Exists(dgst digest.Digest) (bool, error) {
-	// Call by Tag endpoint since the API uses the same
-	// URL endpoint for tags and digests.
-	return ms.ExistsByTag(dgst.String())
+func descriptorFromHeaders(headers http.Header) (distribution.Descriptor, error) {
+	desc := distribution.Descriptor{}
+
+	digestHeader := headers.Get("Docker-Content-Digest")
+	if digestHeader == "" {
+		return distribution.Descriptor{}, errors.New("missing or empty Docker-Content-Digest header")
+	}
+	dgst, err := digest.ParseDigest(digestHeader)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+	desc.Digest = dgst
+
+	lengthHeader := headers.Get("Content-Length")
+	if lengthHeader == "" {
+		return distribution.Descriptor{}, errors.New("missing or empty Content-Length header")
+	}
+	length, err := strconv.ParseInt(lengthHeader, 10, 64)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+	desc.Size = length
+
+	ctHeader := headers.Get("Content-Type")
+	if ctHeader == "" {
+		return distribution.Descriptor{}, errors.New("missing or empty Content=Type header")
+	}
+	desc.MediaType = ctHeader
+	return desc, nil
+
+}
+
+// Get issues a HEAD request for a Manifest against its named endpoint in order
+// to construct a descriptor for the tag.  If the registry doesn't support HEADing
+// a manifest, fallback to GET.
+func (t *tags) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	u, err := t.ub.BuildManifestURL(t.name, tag)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+	var attempts int
+	resp, err := t.client.Head(u)
+
+check:
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	switch {
+	case resp.StatusCode >= 200 && resp.StatusCode < 400:
+		return descriptorFromHeaders(resp.Header)
+	case resp.StatusCode == http.StatusMethodNotAllowed:
+		resp, err = t.client.Get(u)
+		attempts++
+		if attempts > 1 {
+			return distribution.Descriptor{}, err
+		}
+		goto check
+	default:
+		return distribution.Descriptor{}, handleErrorResponse(resp)
+	}
+}
+
+func (t *tags) Lookup(ctx context.Context, digest distribution.Descriptor) ([]string, error) {
```

Comment:
We should be careful about supporting this reverse lookup without more restrictions. I'm guessing this means find all of the tags for objects which have tags pointing at the `digest`. Have we specified that constraint on this interface definition?